File: data_environment/sources/induction.1.c
/*
* @@name:	induction.1
* @@type:	C
* @@operation:	compile
* @@expect:	success
* @@version:	omp_6.0
*/
#include <stdio.h>
#include <math.h>

void comp_poly(int N, double x, double c[]) {
   // x:    input: value of x for which to eval the polynomial
   // c[N]: input: the coefficients
   double x0 = 1.0;       // initial value x^0 == 1
   double xi;             // x^i
   double result;         // accumulator for the result

   // Case 1: induction clause
   xi = x0;
   result = 0.0;
   #pragma omp parallel for reduction(+: result) induction(step(x),*: xi)
   for (int i = 0; i < N; i++) {
      result += c[i] * xi;
      xi *= x;
   }
   printf("C1: result = %f, xn = %f\n", result, xi);

   // Case 2: inscan reduction
   xi = x0;
   result = 0.0;
   #pragma omp parallel for reduction(+: result) reduction(inscan,*: xi)
   for (int i = 0; i < N; i++) {
      result += c[i] * xi;
      #pragma omp scan exclusive(xi)
      xi *= x;
   }
   printf("C2: result = %f, xn = %f\n", result, xi);

   // Case 3: closed form
   result = 0.0;
   #pragma omp parallel for reduction(+: result) lastprivate(xi)
   for (int i = 0; i < N; i++) {
      xi = x0 * pow(x, i);     // induction operation in closed form
      result += c[i] * xi;
      xi *= x;
   }
   printf("C3: result = %f, xn = %f\n", result, xi);
}

File: data_environment/sources/induction.1.f90
! @@name:	induction.1
! @@type:	F-free
! @@operation:	compile
! @@expect:	success
! @@version:	omp_6.0
subroutine comp_poly(N, x, c)
   implicit none
   ! x:    input: value of x for which to eval the polynomial
   ! c(N): input: the coefficients
   integer :: N
   double precision :: x, c(*)

   double precision :: x0 = 1.0  ! initial value x^0 == 1
   double precision :: xi        ! x^i
   double precision :: result    ! accumulator for the result
   integer :: i

   !! Case 1: induction clause
   xi = x0
   result = 0.0
   !$omp parallel do reduction(+: result) induction(step(x),*: xi)
   do i = 1, N
      result = result + c(i) * xi
      xi = xi * x
   end do
   print *, 'C1: result =', result, ', xn =', xi

   !! Case 2: inscan reduction
   xi = x0
   result = 0.0
   !$omp parallel do reduction(+: result) reduction(inscan,*: xi)
   do i = 1, N
      result = result + c(i) * xi
      !$omp scan exclusive(xi)
      xi = xi * x
   end do
   print *, 'C2: result =', result, ', xn =', xi

   !! Case 3: closed form
   result = 0.0
   !$omp parallel do reduction(+: result) lastprivate(xi)
   do i = 1, N
      xi = x0 * (x ** (i-1))      ! induction operation in closed form
      result = result + c(i) * xi
      xi = xi * x
   end do
   print *, 'C3: result =', result, ', xn =', xi
end subroutine
